controlplane/kubeadm/internal/machinefilters/machine_filters.go

```diff
@@ -197,3 +204,140 @@ func ControlPlaneSelectorForCluster(clusterName string) labels.Selector {
+			return true
+		}
+
+		bootstrapObj := &bootstrapv1.KubeadmConfig{}
+		if err := c.Get(ctx, client.ObjectKey{Name: bootstrapRef.Name, Namespace: machine.Namespace}, bootstrapObj); err != nil {
+			// Return true here because failing to get KubeadmConfig should not be considered as unmatching.
+			return true
+		}
+
+		kcpConfigSpecLocal := kcp.Spec.KubeadmConfigSpec.DeepCopy()
+
+		// Machine's init configuration is nil when machine is the control plane is already initialized.
+		if bootstrapObj.Spec.InitConfiguration == nil {
+			kcpConfigSpecLocal.InitConfiguration = nil
+		}
+
+		// Machine's join configuration is nil when a machine is the first machine in the control plane.
+		if bootstrapObj.Spec.JoinConfiguration == nil {
+			kcpConfigSpecLocal.JoinConfiguration = nil
+		} else if kcpConfigSpecLocal.JoinConfiguration == nil {
+			// If KCP join configuration is not present, set machine join configuration to nil (nothing can trigger rollout here).
+			bootstrapObj.Spec.JoinConfiguration = nil
+		}
+
+		if bootstrapObj.Spec.InitConfiguration != nil && kcpConfigSpecLocal.InitConfiguration != nil {
+			bootstrapObj.Spec.InitConfiguration.TypeMeta = kcpConfigSpecLocal.InitConfiguration.TypeMeta
+		}
+
+		if bootstrapObj.Spec.JoinConfiguration != nil && kcpConfigSpecLocal.JoinConfiguration != nil {
+			bootstrapObj.Spec.JoinConfiguration.TypeMeta = kcpConfigSpecLocal.JoinConfiguration.TypeMeta
+		}
+
+		// Machines that have KubeadmClusterConfigurationAnnotation will have to match with KCP ClusterConfiguration.
+		// If the annotation is not present (machine is either old or adopted), we won't roll out on any possible changes
+		// made in KCP's ClusterConfiguration given that we don't have enough information to make a decision.
+		// Users should use KCP.Spec.UpgradeAfter field to force a rollout in this case.
+		machineClusterConfigStr, ok := machine.GetAnnotations()[controlplanev1.KubeadmClusterConfigurationAnnotation]
+		if ok {
+			machineClusterConfig := &kubeadmv1.ClusterConfiguration{}
+			// ClusterConfiguration annotation is not correct, only solution is to rollout.
+			if err := json.Unmarshal([]byte(machineClusterConfigStr), machineClusterConfig); err != nil {
+				return false
+			}
+			if !reflect.DeepEqual(machineClusterConfig, kcp.Spec.KubeadmConfigSpec.ClusterConfiguration) {
+				return false
+			}
+		}
+
+		// KCP ClusterConfiguration will only be compared with a machine's ClusterConfiguration annotation.
+		// To skip ClusterConfiguration merge during SemanticMerge(), set both Machine's and KCP's ClusterConfigurations to nil here.
+		kcpConfigSpecLocal.ClusterConfiguration = nil
+		bootstrapObj.Spec.ClusterConfiguration = nil
+
+		// Machine's JoinConfiguration Discovery and ControlPlane fields are set to empty object by KubeadmConfig controller if nil,
+		// so this field should be removed from Machine's JoinConfiguration if KCP does not have them present.
```

Comment:
Not sure if I fully follow this comment, CABPK usually sets this with a bootstrap token, is that why we're cleaning it up here?

Reply:
KCP discovery/controlplane can be empty. If this is the case, there is nothing to compare, so setting machine's discovery/controlplane to empty. Not just bootstrap token. 

An example of join config on the machines when join config is nil on KCP is:

```
  joinConfiguration:
    controlPlane:
      localAPIEndpoint:
        advertiseAddress: ""
        bindPort: 0
    discovery:
      bootstrapToken:
        apiServerEndpoint: 172.17.0.3:6443
        caCertHashes:
        - sha256:621
        token: pbmxii
        unsafeSkipCAVerification: false
    nodeRegistration:
      criSocket: /var/run/containerd/containerd.sock
      kubeletExtraArgs:
        eviction-hard: nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%

```

Reply:
Btw, when join config on KCP is nil, nodeRegistration on machine matches nodeRegistration on KCP init config.
This is new check I added recently.